File: src/main/java/org/bipmed/brave/vcf/VcfImporter.java
package org.bipmed.brave.vcf;

import htsjdk.variant.variantcontext.Allele;
import htsjdk.variant.variantcontext.Genotype;
import htsjdk.variant.variantcontext.VariantContext;
import htsjdk.variant.vcf.VCFFileReader;
import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;
import org.bipmed.brave.variant.Statistics;
import org.bipmed.brave.variant.Variant;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

import java.io.File;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

@Component
public class VcfImporter {

    private RestTemplate restTemplate;

    @Autowired
    public VcfImporter(RestTemplate restTemplate) {
        this.restTemplate = restTemplate;
    }

    public long importVariants(String filename, String datasetId, String assemblyId) {
        VCFFileReader vcfFileReader = new VCFFileReader(new File(filename), false);

        long count = 0L;

        for (VariantContext variantContext : vcfFileReader) {
            DescriptiveStatistics coverages = new DescriptiveStatistics(variantContext.getGenotypes().stream().mapToDouble(Genotype::getDP).toArray());
            DescriptiveStatistics genotypeQualities = new DescriptiveStatistics(variantContext.getGenotypes().stream().mapToDouble(Genotype::getGQ).toArray());

            Variant variant = Variant.builder()
                    .snpIds(!variantContext.getID().equals(".") ? Arrays.asList(variantContext.getID().split(";")) : Collections.emptyList())
                    .datasetId(datasetId)
                    .assemblyId(assemblyId)
                    .totalSamples(vcfFileReader.getFileHeader().getNGenotypeSamples())
                    .referenceName(variantContext.getContig().replaceFirst("^chr", ""))
                    .geneSymbol(getAnnotationColumn(variantContext, 3))
                    .sampleCount(Long.parseLong(variantContext.getAttributeAsString("NS", null)))
                    .start((long) variantContext.getStart())
                    .alleleFrequency(variantContext.getAttributeAsStringList("AF", null).stream().map(Float::parseFloat).collect(Collectors.toList()))
                    .alternateBases(variantContext.getAlternateAlleles().stream().map(Allele::getBaseString).collect(Collectors.toList()))
                    .referenceBases(variantContext.getReference().getBaseString())

                    .coverage(Statistics.builder()
                            .min(coverages.getMin())
                            .q25(coverages.getPercentile(25))
                            .median(coverages.getPercentile(50))
                            .q75(coverages.getPercentile(75))
                            .max(coverages.getMax())
                            .mean(coverages.getMean())
                            .build())

                    .genotypeQuality(Statistics.builder()
                            .min(genotypeQualities.getMin())
                            .q25(genotypeQualities.getPercentile(25))
                            .median(genotypeQualities.getPercentile(50))
                            .q75(genotypeQualities.getPercentile(75))
                            .max(genotypeQualities.getMax())
                            .mean(genotypeQualities.getMean())
                            .build())

                    .clnsig(variantContext.getAttributeAsString("CLNSIG", null))
                    .hgvs(getAnnotationColumn(variantContext, 9))
                    .type(getAnnotationColumn(variantContext, 5))
                    .build();

            try {
                restTemplate.postForLocation("/variants", variant);
            } catch (Exception e) {
                System.out.println("ERROR: " + e.getMessage());
                System.exit(1);

            }

            count++;
        }

        return count;
    }

    private List<String> getAnnotationColumn(VariantContext variant, Integer index) {
        if (variant.hasAttribute("ANN")) {
            return Arrays.stream(variant.getAttribute("ANN").toString().split(","))
                    .map(it -> it.split("\\|")[index])
                    .collect(Collectors.toList());
        } else {
            return null;
        }
    }
}

File: src/main/kotlin/org/bipmed/import/vcf/VcfImporter.kt
package org.bipmed.import.vcf

import htsjdk.variant.variantcontext.VariantContext
import htsjdk.variant.vcf.VCFFileReader
import org.bipmed.import.variant.Variant
import org.nield.kotlinstatistics.median
import org.nield.kotlinstatistics.percentile
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.stereotype.Component
import java.io.File

@Component
class VcfImporter(private val mongoTemplate: MongoTemplate) {


    fun import(filename: String, datasetId: String, assemblyId: String): Long {
        val vcfFileReader = VCFFileReader(File(filename), false)

        var count = 0L

        vcfFileReader.forEach {
            val coverages = it.genotypes.map { genotype -> genotype.dp }
            val genotypeQualities = it.genotypes.map { genotype -> genotype.gq }

            val variant = Variant(
                    snpIds = if (it.id != ".") it.id.split(';') else emptyList(),
                    geneSymbol = getGeneSymbol(it),
                    sampleCount = it.getAttributeAsString("NS", null)?.toLong(),
                    start = it.start.toLong(),
                    referenceName = it.contig.removePrefix("chr"),
                    datasetId = datasetId,
                    alleleFrequency = it.getAttributeAsStringList("AF", null).map { af -> af.toDouble() },
                    alternateBases = it.alternateAlleles.map { allele -> allele.baseString },
                    referenceBases = it.reference.baseString,
                    assemblyId = assemblyId,

                    coverage = Variant.Statistics(
                            min = coverages.min(),
                            q25 = coverages.percentile(25.0),
                            median = coverages.median(),
                            q75 = coverages.percentile(75.0),
                            max = coverages.max(),
                            mean = coverages.average()
                    ),

                    genotypeQuality = Variant.Statistics(
                            min = genotypeQualities.min(),
                            q25 = genotypeQualities.percentile(25.0),
                            median = genotypeQualities.median(),
                            q75 = genotypeQualities.percentile(75.0),
                            max = genotypeQualities.max(),
                            mean = genotypeQualities.average()
                    ),

                    clnsig = it.getAttributeAsString("CLNSIG", null)
            )

            mongoTemplate.insert(variant)
            count++
        }

        return count
    }

    private fun getGeneSymbol(variant: VariantContext): String? {
        return if (variant.hasAttribute("ANN")) {
            variant.getAttribute("ANN").toString().split('|')[3] // Gene Name
        } else {
            null
        }
    }

}